File: src/main/java/xyz/block/veeseven/UuidV7.java
package xyz.block.veeseven;

import java.security.SecureRandom;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.LongSupplier;

/**
 * Utility class for generating and working with UUID v7 identifiers.
 * <p>
 * UUID v7 is a time-ordered UUID format that encodes a Unix timestamp in milliseconds
 * in the most significant 48 bits, making UUIDs naturally sortable by creation time.
 * This implementation follows RFC 9562 with optional monotonic counter support.
 * <p>
 * Non-monotonic methods use ThreadLocalRandom for maximum performance with no synchronization.
 * Monotonic methods use a synchronized counter to ensure strict ordering within the same millisecond.
 */
public final class UuidV7 {

    // Counter occupies 12 bits (rand_a in RFC 9562)
    private static final int COUNTER_BITS = 12;
    private static final int COUNTER_MAX = (1 << COUNTER_BITS) - 1; // 0xFFF

    // Thread-safe random source for initial counter values in monotonic mode
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    // Monotonicity state (guarded by class lock in monotonic methods)
    private static long monotonicGeneratorLastTimestamp = 0L;
    private static int monotonicGeneratorCounter = 0;

    private UuidV7() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }

    /**
     * Generates a new UUID v7 using the current system time in non-monotonic mode.
     * <p>
     * Uses ThreadLocalRandom for maximum performance with no synchronization overhead.
     * UUIDs generated in the same millisecond may not be strictly ordered, but uniqueness
     * is guaranteed through random bits.
     *
     * @return a new UUID v7 instance
     */
    public static UUID generate() {
        return generateNonMonotonic(System::currentTimeMillis);
    }

    /**
     * Generates a new UUID v7 using a custom clock source in non-monotonic mode.
     * <p>
     * Uses ThreadLocalRandom for maximum performance with no synchronization overhead.
     * Useful for testing or specialized use cases where you need control over the timestamp.
     *
     * @param clock a supplier that returns the current time in milliseconds since Unix epoch
     * @return a new UUID v7 instance
     */
    public static UUID generate(LongSupplier clock) {
        return generateNonMonotonic(clock);
    }

    /**
     * Non-monotonic implementation - no synchronization for maximum performance.
     */
    private static UUID generateNonMonotonic(LongSupplier clock) {
        long timestamp = clock.getAsLong();

        // Use random bits for counter field (rand_a)
        int counterValue = ThreadLocalRandom.current().nextInt(COUNTER_MAX + 1);

        // Generate random bytes for the least significant bits (rand_b)
        byte[] randomBytes = new byte[8];
        ThreadLocalRandom.current().nextBytes(randomBytes);

        return buildUuid(timestamp, counterValue, randomBytes);
    }

    /**
     * Generates a new UUID v7 using the current system time with monotonic ordering.
     * <p>
     * This method ensures that UUIDs generated within the same millisecond are strictly
     * ordered by incrementing a counter. If the counter overflows within a millisecond,
     * the method will block until the next millisecond to maintain uniqueness.
     * <p>
     * This method is synchronized and best suited for database primary keys and scenarios
     * requiring guaranteed sequential ordering.
     *
     * @return a new UUID v7 instance
     */
    public static synchronized UUID generateMonotonic() {
        return generateMonotonicImpl(System::currentTimeMillis);
    }

    /**
     * Generates a new UUID v7 using a custom clock source with monotonic ordering.
     * <p>
     * This method ensures that UUIDs generated within the same millisecond are strictly
     * ordered by incrementing a counter. Useful for testing monotonic behavior with
     * controlled clock sources.
     * <p>
     * This method is synchronized and best suited for database primary keys and scenarios
     * requiring guaranteed sequential ordering.
     *
     * @param clock a supplier that returns the current time in milliseconds since Unix epoch
     * @return a new UUID v7 instance
     */
    public static synchronized UUID generateMonotonic(LongSupplier clock) {
        return generateMonotonicImpl(clock);
    }

    /**
     * Monotonic implementation - synchronized to ensure ordering.
     * This method must only be called from synchronized methods.
     */
    private static UUID generateMonotonicImpl(LongSupplier clock) {
        long timestamp = clock.getAsLong();
        int counterValue;

        if (timestamp == monotonicGeneratorLastTimestamp) {
            // Same millisecond - increment counter
            monotonicGeneratorCounter = (monotonicGeneratorCounter + 1) & COUNTER_MAX;

            if (monotonicGeneratorCounter == 0) {
                // Counter overflow - wait for next millisecond to maintain uniqueness
                do {
                    timestamp = clock.getAsLong();
                } while (timestamp == monotonicGeneratorLastTimestamp);

                // New millisecond - start with random counter value
                monotonicGeneratorCounter = SECURE_RANDOM.nextInt(COUNTER_MAX + 1);
            }
            counterValue = monotonicGeneratorCounter;
        } else {
            // New millisecond - start with random counter value for unpredictability
            monotonicGeneratorCounter = SECURE_RANDOM.nextInt(COUNTER_MAX + 1);
            counterValue = monotonicGeneratorCounter;
            monotonicGeneratorLastTimestamp = timestamp;
        }

        // Generate random bytes for the least significant bits (rand_b)
        byte[] randomBytes = new byte[8];
        ThreadLocalRandom.current().nextBytes(randomBytes);

        return buildUuid(timestamp, counterValue, randomBytes);
    }

    /**
     * Builds a UUID v7 from timestamp, counter, and random bytes.
     */
    private static UUID buildUuid(long timestamp, int counterValue, byte[] randomBytes) {
        // Layout of UUID v7 (RFC 9562):
        //
        // Most significant 64 bits:
        //   48 bits: unix_ts_ms (timestamp in milliseconds)
        //   4 bits:  ver (version = 7)
        //   12 bits: rand_a (counter for monotonicity or random)
        //
        // Least significant 64 bits:
        //   2 bits:  var (variant = 10)
        //   62 bits: rand_b (random data)

        long mostSigBits = (timestamp << 16) | counterValue;

        // Set version to 7 (0111 in bits 48-51)
        mostSigBits = (mostSigBits & 0xFFFFFFFFFFFF0FFFL) | 0x0000000000007000L;

        // Build least significant bits from random bytes
        long leastSigBits = 0L;
        for (int i = 0; i < 8; i++) {
            leastSigBits = (leastSigBits << 8) | (randomBytes[i] & 0xFFL);
        }

        // Set variant to 10 (RFC 4122) in bits 64-65
        leastSigBits = (leastSigBits & 0x3FFFFFFFFFFFFFFFL) | 0x8000000000000000L;

        return new UUID(mostSigBits, leastSigBits);
    }

    /**
     * Extracts the timestamp component from a UUID v7.
     *
     * @param uuid the UUID v7 to extract the timestamp from
     * @return the timestamp in milliseconds since Unix epoch
     * @throws IllegalArgumentException if uuid is null or not a v7 UUID
     */
    public static long getTimestamp(UUID uuid) {
        if (uuid == null) {
            throw new IllegalArgumentException("UUID cannot be null");
        }

        if (uuid.version() != 7) {
            throw new IllegalArgumentException("UUID is not version 7 (got version " + uuid.version() + ")");
        }

        // Extract 48-bit timestamp from most significant bits
        return uuid.getMostSignificantBits() >>> 16;
    }
}

File: swift/Sources/UUIDv7/UUIDv7.swift
import Foundation

/// Utility for generating and working with UUID v7 identifiers.
///
/// UUID v7 is a time-ordered UUID format that encodes a Unix timestamp in milliseconds
/// in the most significant 48 bits, making UUIDs naturally sortable by creation time.
/// This implementation follows RFC 9562.
///
/// This struct uses non-cryptographic random for maximum performance with no synchronization overhead.
/// UUIDs generated in the same millisecond may not be strictly ordered, but uniqueness
/// is guaranteed through random bits. For monotonic ordering guarantees, use
/// `MonotonicUUIDv7` instead.
public enum UUIDv7 {
    
    public typealias Clock = () -> UInt64
    
    /// Generates a new UUID v7 using the current system time.
    ///
    /// Uses non-cryptographic random for maximum performance with no synchronization overhead.
    /// UUIDs generated in the same millisecond may not be strictly ordered, but uniqueness
    /// is guaranteed through random bits.
    ///
    /// - Returns: A new UUID v7 instance
    public static func generate() -> UUID {
        generate(clock: systemClock)
    }
    
    /// Generates a new UUID v7 using a custom clock source.
    ///
    /// Uses non-cryptographic random for maximum performance with no synchronization overhead.
    /// Useful for testing or specialized use cases where you need control over the timestamp.
    ///
    /// - Parameter clock: A closure that returns the current time in milliseconds since Unix epoch
    /// - Returns: A new UUID v7 instance
    public static func generate(clock: Clock) -> UUID {
        let timestamp = clock()
        let randA = UInt16.random(in: 0...0xFFF)
        let randB = UInt64.random(in: 0...UInt64.max)
        return build(timestamp: timestamp, randA: randA, randB: randB)
    }
    
    /// Generates a new UUID v7 as a compact string using the current system time.
    ///
    /// Equivalent to calling `toCompactString(generate())`.
    /// Returns a 22-character Base62 encoded string that preserves time-ordering.
    ///
    /// - Returns: A 22-character compact string representation of a new UUID v7
    public static func generateCompactString() -> String {
        toCompactString(generate())
    }
    
    /// Generates a new UUID v7 as a compact string using a custom clock source.
    ///
    /// Equivalent to calling `toCompactString(generate(clock:))`.
    /// Returns a 22-character Base62 encoded string that preserves time-ordering.
    ///
    /// - Parameter clock: A closure that returns the current time in milliseconds since Unix epoch
    /// - Returns: A 22-character compact string representation of a new UUID v7
    public static func generateCompactString(clock: Clock) -> String {
        toCompactString(generate(clock: clock))
    }
    
    /// Builds a UUID v7 from timestamp and random components.
    ///
    /// - Parameters:
    ///   - timestamp: The timestamp in milliseconds since Unix epoch
    ///   - randA: The random or counter value for bits 52-63 (12 bits)
    ///   - randB: The random value for bits 66-127 (62 bits, variant will be set)
    /// - Returns: A new UUID v7 instance
    static func build(timestamp: UInt64, randA: UInt16, randB: UInt64) -> UUID {
        let maskedRandA = UInt64(randA & 0xFFF)
        var mostSigBits = (timestamp << 16) | maskedRandA
        var leastSigBits = randB
        
        // Set version to 7 (0111 in bits 48-51)
        mostSigBits = (mostSigBits & 0xFFFFFFFFFFFF0FFF) | 0x0000000000007000
        
        // Set variant to 10 (RFC 4122) in bits 64-65
        leastSigBits = (leastSigBits & 0x3FFFFFFFFFFFFFFF) | 0x8000000000000000
        
        return uuidFromBits(mostSigBits: mostSigBits, leastSigBits: leastSigBits)
    }
    
    /// Extracts the timestamp component from a UUID v7.
    ///
    /// - Parameter uuid: The UUID v7 to extract the timestamp from
    /// - Returns: The timestamp in milliseconds since Unix epoch
    /// - Throws: `UUIDv7Error.notVersion7` if the UUID is not version 7
    public static func getTimestamp(_ uuid: UUID) throws -> UInt64 {
        let (mostSigBits, _) = bitsFromUUID(uuid)
        
        let version = Int((mostSigBits >> 12) & 0xF)
        guard version == 7 else {
            throw UUIDv7Error.notVersion7(actualVersion: version)
        }
        
        return mostSigBits >> 16
    }
    
    /// Default system clock returning milliseconds since Unix epoch.
    public static let systemClock: Clock = {
        UInt64(Date().timeIntervalSince1970 * 1000)
    }
}

// MARK: - Compact String Encoding

extension UUIDv7 {
    
    private static let base62Alphabet = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
    private static let base62Length = 22
    
    /// Converts a UUID to a compact string representation.
    ///
    /// The resulting string is exactly 22 characters long and preserves lexicographic
    /// ordering for UUID v7 values (time-ordered UUIDs will sort correctly as compact strings).
    /// Uses Base62 encoding (0-9, A-Z, a-z).
    ///
    /// - Parameter uuid: The UUID to encode
    /// - Returns: A 22-character compact string representation
    public static func toCompactString(_ uuid: UUID) -> String {
        let (mostSigBits, leastSigBits) = bitsFromUUID(uuid)
        
        // Convert to two UInt64s and then to base62
        var high = mostSigBits
        var low = leastSigBits
        
        var result = [Character](repeating: "0", count: base62Length)
        
        // Process from least significant digit
        for i in stride(from: base62Length - 1, through: 0, by: -1) {
            // Divide 128-bit number by 62
            let (newHigh, newLow, remainder) = divide128By62(high: high, low: low)
            result[i] = base62Alphabet[Int(remainder)]
            high = newHigh
            low = newLow
        }
        
        return String(result)
    }
    
    /// Decodes a compact string representation back to a UUID.
    ///
    /// - Parameter compactString: The compact string to decode (must be 22 characters)
    /// - Returns: The decoded UUID
    /// - Throws: `UUIDv7Error` if the string is invalid
    public static func fromCompactString(_ compactString: String) throws -> UUID {
        guard compactString.count == base62Length else {
            throw UUIDv7Error.invalidCompactStringLength(length: compactString.count)
        }
        
        var high: UInt64 = 0
        var low: UInt64 = 0
        
        for char in compactString {
            guard let digit = base62Value(of: char) else {
                throw UUIDv7Error.invalidCompactStringCharacter(character: char)
            }
            
            // Multiply 128-bit number by 62 and add digit
            (high, low) = multiply128By62AndAdd(high: high, low: low, add: UInt64(digit))
        }
        
        return uuidFromBits(mostSigBits: high, leastSigBits: low)
    }
    
    private static func base62Value(of char: Character) -> Int? {
        switch char {
        case "0"..."9":
            return Int(char.asciiValue! - Character("0").asciiValue!)
        case "A"..."Z":
            return Int(char.asciiValue! - Character("A").asciiValue!) + 10
        case "a"..."z":
            return Int(char.asciiValue! - Character("a").asciiValue!) + 36
        default:
            return nil
        }
    }
    
    private static func divide128By62(high: UInt64, low: UInt64) -> (high: UInt64, low: UInt64, remainder: UInt64) {
        // Divide 128-bit number by 62 using long division
        let divisor: UInt64 = 62
        
        // Divide high part
        let highQuotient = high / divisor
        let highRemainder = high % divisor
        
        // Combine remainder with low part
        // (highRemainder << 64) + low, divide by 62
        // Since highRemainder < 62, we can use a simpler approach
        
        let (lowQuotient, lowRemainder) = divideWithCarry(
            high: highRemainder,
            low: low,
            divisor: divisor
        )
        
        return (highQuotient, lowQuotient, lowRemainder)
    }
    
    private static func divideWithCarry(high: UInt64, low: UInt64, divisor: UInt64) -> (quotient: UInt64, remainder: UInt64) {
        // Split low into two 32-bit parts for easier handling
        let lowHigh = low >> 32
        let lowLow = low & 0xFFFFFFFF
        
        // Combine high remainder with upper 32 bits of low
        let upper = (high << 32) | lowHigh
        let upperQuotient = upper / divisor
        let upperRemainder = upper % divisor
        
        // Combine upper remainder with lower 32 bits
        let lower = (upperRemainder << 32) | lowLow
        let lowerQuotient = lower / divisor
        let lowerRemainder = lower % divisor
        
        // Combine quotients
        let quotient = (upperQuotient << 32) | lowerQuotient
        
        return (quotient, lowerRemainder)
    }
    
    private static func multiply128By62AndAdd(high: UInt64, low: UInt64, add: UInt64) -> (high: UInt64, low: UInt64) {
        let multiplier: UInt64 = 62
        
        // Multiply low part
        let (lowProduct, lowOverflow) = low.multipliedReportingOverflow(by: multiplier)
        var lowCarry: UInt64 = 0
        if lowOverflow {
            // Calculate the high part of the multiplication manually
            lowCarry = low.multipliedFullWidth(by: multiplier).high
        }
        
        // Multiply high part and add carry from low
        let (highProduct, _) = high.multipliedReportingOverflow(by: multiplier)
        let newHigh = highProduct &+ lowCarry
        
        // Add the digit
        let (newLow, addOverflow) = lowProduct.addingReportingOverflow(add)
        let finalHigh = addOverflow ? newHigh &+ 1 : newHigh
        
        return (finalHigh, newLow)
    }
}

// MARK: - UUID Bit Manipulation

extension UUIDv7 {
    
    static func bitsFromUUID(_ uuid: UUID) -> (mostSigBits: UInt64, leastSigBits: UInt64) {
        var bytes = uuid.uuid
        
        // UUID bytes are stored in big-endian order
        let mostSigBits = withUnsafeBytes(of: &bytes) { ptr -> UInt64 in
            var value: UInt64 = 0
            for i in 0..<8 {
                value = (value << 8) | UInt64(ptr[i])
            }
            return value
        }
        
        let leastSigBits = withUnsafeBytes(of: &bytes) { ptr -> UInt64 in
            var value: UInt64 = 0
            for i in 8..<16 {
                value = (value << 8) | UInt64(ptr[i])
            }
            return value
        }
        
        return (mostSigBits, leastSigBits)
    }
    
    static func uuidFromBits(mostSigBits: UInt64, leastSigBits: UInt64) -> UUID {
        var bytes = (UInt8(0), UInt8(0), UInt8(0), UInt8(0),
                     UInt8(0), UInt8(0), UInt8(0), UInt8(0),
                     UInt8(0), UInt8(0), UInt8(0), UInt8(0),
                     UInt8(0), UInt8(0), UInt8(0), UInt8(0))
        
        // Store in big-endian order
        bytes.0 = UInt8((mostSigBits >> 56) & 0xFF)
        bytes.1 = UInt8((mostSigBits >> 48) & 0xFF)
        bytes.2 = UInt8((mostSigBits >> 40) & 0xFF)
        bytes.3 = UInt8((mostSigBits >> 32) & 0xFF)
        bytes.4 = UInt8((mostSigBits >> 24) & 0xFF)
        bytes.5 = UInt8((mostSigBits >> 16) & 0xFF)
        bytes.6 = UInt8((mostSigBits >> 8) & 0xFF)
        bytes.7 = UInt8(mostSigBits & 0xFF)
        
        bytes.8 = UInt8((leastSigBits >> 56) & 0xFF)
        bytes.9 = UInt8((leastSigBits >> 48) & 0xFF)
        bytes.10 = UInt8((leastSigBits >> 40) & 0xFF)
        bytes.11 = UInt8((leastSigBits >> 32) & 0xFF)
        bytes.12 = UInt8((leastSigBits >> 24) & 0xFF)
        bytes.13 = UInt8((leastSigBits >> 16) & 0xFF)
        bytes.14 = UInt8((leastSigBits >> 8) & 0xFF)
        bytes.15 = UInt8(leastSigBits & 0xFF)
        
        return UUID(uuid: bytes)
    }
}

// MARK: - Errors

public enum UUIDv7Error: Error, Equatable {
    case notVersion7(actualVersion: Int)
    case invalidCompactStringLength(length: Int)
    case invalidCompactStringCharacter(character: Character)
}

extension UUIDv7Error: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .notVersion7(let actualVersion):
            return "UUID is not version 7 (got version \(actualVersion))"
        case .invalidCompactStringLength(let length):
            return "Compact string must be exactly 22 characters (got \(length))"
        case .invalidCompactStringCharacter(let character):
            return "Invalid compact string character: \(character)"
        }
    }
}
